### PR TITLE
73x times faster metrics correlations at the agent

### PR DESCRIPTION
**Pull request by ktsaou**
This PR improves the performance of metrics correlations by a factor of 73.

The biggest problem of it was this function:

```c
int find_index(double arr[], long int n, double K, long int start)
{
    for (long int i = start; i < n; i++) {
        if (K<arr[i]){
            return i;
        }
    }
    return n;
}
```

It has been replaced with a binary search function, which alone saved 75% of CPU resources required.

Then the next big problem was comparisons of `double` numbers. So, I multiplied the `double` pair diffs by 1 million (to have 6 digits precision) and divided the final `min` and `max` by 1 million, so that we only compare `long int` numbers. This saved 15% more (compared to the original).

Then a number of additional optimizations have been done to optimize the whole logic further:

1. The need to replace NANs with zeros has been replaced with the option `RRDR_OPTION_NULL2ZERO` so that the query engine will do it.
2. Loops to copy data around have been eliminated
3. Loops to calculate min and max have been eliminated

The current code in the master, has another important issue. The baseline is 4x bigger than the highlight, but the code requests the same number of points for both the highlight and the baseline. This is now fixed. The new code tries to have the same duration for each point of the result. So, it first does the highlight query to find the number of points returned and then it does the baseline query requesting 4 times the number of points of the highlight. The code is smart enough to adapt the baseline window to be power of 2 compared to the highlight and respecting the maximum buffers the code has. So, the old code, also gives wrong results.

Overall, for a given test I made for 30 minutes of metrics correlations, by observations are like follows:

1. The original time to run metric correlations was about 8 seconds.
2. The time to do the queries to collect the data was about 700 ms (just 9% of the total time).
3. The new time to run metric correlations is about 800 ms (for the same timeframe, the same dataset).
4. So, 7300ms before vs 100ms now. 73x times faster.

Now metrics correlations is just about 12% more resource hungry than running metrics correlations on the cloud. It is really nothing. If we accumulated into this the CPU time the agent needs to prepare JSON responses for all data queries and send them over to the cloud, the agent version is now really a lot more efficient for the agent itself.

I also enabled metric correlations at the agent, by default.

### CODE REVIEW

**Comment by ktsaou** [comment]
@andrewm4894 you have to review this.

I found some discrepancies with the new and the old logic.
For example, the previous logic was giving about 2500 dimensions with a value of 1. Now the new logic gives only 250. All the others are the same. Just a lot of 1s became 0s. I really can't tell why.

Still this logic needs to be validated.

**Comment by andrewm4894** [comment]
Good stuff! I'll spin up some agents on this branch and do same comparison to the cloud results we did originally.

**Comment by ilyam8** [comment]
@andrewm4894 @MrZammler If that is technically possible I'd suggest adding some unit tests with input and **expected** correlations output. So we can improve our implementation and not be afraid that something is broken.

**Comment by ktsaou** [comment]
Now its 10x times faster!

**Comment by ktsaou** [comment]
So, the actual comparison is like this:

1. The queries need 700ms
2. The original time was 8 seconds. So 7300ms was metric correlations
3. Now total time is 800ms. Minus the queries (700ms), metric correlations is now 100ms.
4. So, the improvement is like 73 times faster...

**Comment by andrewm4894** [comment]
> @andrewm4894 @MrZammler If that is technically possible I'd suggest adding some unit tests with input and **expected** correlations output. So we can improve our implementation and not be afraid that something is broken.

Yeah it's a bit manual at the moment since have to test against cloud but going forward for sure we could have test data and expected outputs just for the endpoint which would be easy when testing iterations of the api and not needing to validate against cloud since we really only needed to do that for the original agent implementation. Will follow up with @MrZammler to make separate issues for some unit tests like you mention.

**Comment by ktsaou** [comment]
Guys, I updated the code. So please try the new one.
I think it is done now. I am testing it too.

**Comment by ktsaou** [comment]
Difference on the agent compared to the cloud version, using a 48 minute highlighted area, running on RPI4b.

## Metric Correlations of this PR:

image

## Metric Correlations of cloud version (impact on agent):

image

As you can see, the CPU load is twice as high (2 cores used out of 4) and 13 seconds instead of 8 of the version above.
And the cloud does NOT work - does not bring results (I think it times out) !

**Comment by andrewm4894** [comment]
some initial testing of resources on identical machines, one on master one running this branch. 

I highlighted a 10 min window and ran MC on both agents. 

yellow is what's on master, blue is this branch.

image

i will test that the actual results returned are the same between both branches and vs cloud

**Comment by andrewm4894** [comment]
hmm - so if i use [this notebook](https://colab.research.google.com/drive/1DHx7_23JPQyZwLc-VtSmeKFirvJWjHD_?usp=sharing) to compare agent to cloud based on a node running on master branch i see this:

image

which is very very close between agent to cloud results

But if i do same process for a node on this branch i see this:

image

Which is not as good a match and seems to show the agent having a bias towards a lower p-value. 

@MrZammler this is feeling a little similar to what we saw the first time around - any thoughts?

**Comment by andrewm4894** [comment]
@ktsaou is there any way to version this endpoint somehow such that would be easier to test different algo "generations" on the same agent. Might be a lot of extra work though maybe.

Basically we need to validate the results returned against whats actually live on master and ideally being able to just do this on same node would be best. 

Or if we just had a static range of test input datasets and expected outputs like @ilyam8 suggested that would also maybe be best way forward.

Or for now we can just use the `run_on` param that we can pass to cloud to tell it to run on `cloud` or `agent` and just keep comparing the `agent` results here to `cloud` but its a bit slower and painful this way.

**Comment by andrewm4894** [comment]
maybe even moving this stuff to something like `api/v1/metric_correlations_tmp` so that we could compare results of `api/v1/metric_correlations` to `api/v1/metric_correlations_tmp` on the same agent. 

But no idea if such an approach would be impossible given the implementation details so i'm not sure on best approach.

**Comment by andrewm4894** [comment]
anyway its easy enough to just repeat the notebook scatter plot approach each time we have changes to test on this branch

**Comment by ktsaou** [comment]
@andrewm4894 something is wrong here. The algorithm currently on master, requests the same number of points for both the baseline and the highlight.

Since you are saying that the cloud matches closes the algorithm currently on master, I understand that BOTH of them have the same problem: The baseline is 4x the window but the number of points are the same!

On my version, I fixed that. The baseline has 4 times more points than the hightlight, so that the duration of each point remains the same.

Other than that, the algorithm should be the same.

**Comment by andrewm4894** [comment]
After discussion with @ktsaou i am in agreement that what we have now is probably more robust to certain edge cases than even our existing cloud implementation (the part that gets the data to the python svc) such that my scatter plot approach above is not as valid as i had thought. 

So after qualitative testing of this i am seeing good results (and maybe even less false positives, but just a feeling) and so happy that we should merge this.

**Review by andrewm4894 — APPROVED** [review]

**Comment by andrewm4894** [comment]
actually - one last test I want to do is validate things a little using [this notebook](https://colab.research.google.com/drive/1PH0loJvWJFCgqD0vzbI2RzQdvzlfu1ck?usp=sharing) which basically calculates `p_api`, `p_py`, `p_staging` and compares them all.

**Comment by andrewm4894** [comment]
from [this notebook](https://colab.research.google.com/drive/1PH0loJvWJFCgqD0vzbI2RzQdvzlfu1ck?usp=sharing)

`p_py` vs `p_cloud` seems to match very well:

image

whereas `p_py` vs `p_agent` still matches ok but not as well: 

image

But tbh its close enough and probalby just accounts for some precision we have traded off for performance on with this branch. 

I am happy to merge. will approve.

**Review by andrewm4894 — DISMISSED** [review]

**Comment by ktsaou** [comment]
Valgrind is reporting that there are some uninitialized data that are used in kstwo. Like not filling the arrays completely. I will try to understand what is happenning tomorrow.

Btw, I think the current master also has a memory leak which I fixed. 

Please don't merge until I say so.

**Comment by thiagoftsm** [comment]
@ktsaou when you feel it is ready, please, let us know, because I will review and test your code. :handshake:

**Comment by ktsaou** [comment]
The previous code (currently in master) has 2 additional issues:

### Memory leak
This allocation is never freed:
https://github.com/netdata/netdata/blob/6e3144683a73a2024d51425b20ecfd569034c858/database/metric_correlations.c#L154

### not computed dimensions were included in the results

This loop:
https://github.com/netdata/netdata/blob/6e3144683a73a2024d51425b20ecfd569034c858/database/metric_correlations.c#L150

And this loop:
https://github.com/netdata/netdata/blob/6e3144683a73a2024d51425b20ecfd569034c858/database/metric_correlations.c#L186

Both loops do not check if the dimensions were actually selected by the query engine. When dimensions were not selected by the query engine the `kstwo` algorithm was actually run against random memory garbage that happened to be at the position of these dimensions in the results.

**Comment by ktsaou** [comment]
### About memory efficiency

The code currently in master, allocates the following for every dimension in a chart:

https://github.com/netdata/netdata/blob/6e3144683a73a2024d51425b20ecfd569034c858/database/metric_correlations.c#L15-L22

`MAX_POINTS` is 10k.

`calculated_number` is 16 bytes and `double` is 8 bytes.
So, for every dimension it needed:

- 16 x 10k
- 16 x 10k
- 8 x 10k
- 8 x 10k

Total = 480k for every single dimension. `apps.cpu` with 30 dimensions would need about 15MB. And this is not the peak. At peak it needed that with 1 of the two RRDR results loaded!

The new code loads the 2 RRDRs together and then for every dimension it uses on the stack 4 arrays like above, but with a size equal to the actual points needed. Only for one dimension at a time and only for the number of points returned by the query.

**Comment by ktsaou** [comment]
### About locks efficiency

https://github.com/netdata/netdata/blob/6e3144683a73a2024d51425b20ecfd569034c858/database/metric_correlations.c#L247-L260

The code currently in master locks ALL charts for the entire time metrics correlations is running. This means that until metrics correlations finishes, no new dimensions can be added to any of the charts of the host.

The new code, locks every chart only for the time it is being queried.

**Comment by ktsaou** [comment]
### HTTP response code

The code currently in master, always returns HTTP/200 OK, even if it cannot do the job or it rejects the request.

The new code returns:

- HTTP/200 OK, when the work has been done and there is at least 1 correlated dimension
- HTTP/400 Bad Request, when the parameters given are invalid and do not make sense
- HTTP/403 Forbidden, when metric correlations are not enabled at the agent
- HTTP/404 Not Found, when there are no charts, or no results have been produced by metric correlations
- HTTP/504 Gateway Timeout, when the request timed out (it now accepts a `&timeout=MS` parameter)

**Comment by ktsaou** [comment]
Guys, this is ready for review.
I don't have any other changes to make.

**Comment by ktsaou** [comment]
Found out that for big durations, the query was done without the `unaligned` option. This had the effect that the query was shifted to the left or the right and was not exactly matching the highlighted area.

Also, I saw that we were querying even the dimensions that are just zero for both the baseline and the highlighted windows. Now the code does not evaluate them (skips them).

**Comment by ktsaou** [comment]
So, I was showing this algorithm to my son and at some point he says: "hey dad, I think there is simpler way to do all these. Why don't you calculate the volume of the baseline window and the volume of the highlighted window, normalize them to be comparable and just find the percentage they are different."

Hm... Nice idea... The following branch has that code: https://github.com/ktsaou/netdata/tree/metrics-correlations-with-volume

It works like this:

1. Calculate the volume of the baseline
2. Calculate the volume of the highlight
3. Since the baseline is bigger, divide the baseline with the right number to be comparable to the highlight
4. Then calculate the percentage they are different, using `(base_volume - high_volume) / base_volume`. This produces a number from zero to whatever, based on how different they are. Exception: if `base_volume  = 0` this calculation gives infinite, so we assume a very high percentage (1000000000).
6. Then get all these numbers for all the dimensions into an array.
7. Sort that array.
8. Remove the duplicate values from it.
9. So X unique values are now in the array.
10. Diving 1 with X to find the weight of each slot.
11. Then loop through all dimensions to find their slot in the sorted and unique array.
12. Multiply their slot number with the weight
13. Return that number back to the UI

If you install it and try it:

- the far left of the slider has only the dimensions that their baseline was zero but in the highlighted window they had values.
- the far right of the slider shows all the dimensions, even the ones that are just constant over the whole time.
- any other position of the slider proportionally selects less or more dimensions based on how much different is their highlighted window from their baseline window.

Nice idea from a 17 year old boy...

**Comment by ktsaou** [comment]
I merged the volume implementation into this.
Now the rest API for metric correlations supports a `method=ks2` or `method=volume`. If none is given, `ks2` is assumed.

I changed the `volume` method, to actually query each dimension individually instead of doing chart queries. A lot more queries, but a lot faster too.

`max_points` is irrelevant for `volume` calculations. It actually calculates only the average over the baseline and the highlight.

**Comment by ktsaou** [comment]
@andrewm4894 volume based correlations are very easy to be done across nodes. The agent will return 2 averages for each dimension (baseline and highlight) and the cloud will aggregate these across nodes and do the post processing (sort them and find the right weight for each of them at a multi-node level).

**Comment by ktsaou** [comment]
@andrewm4894 the latest commit (spread results evenly) re-distributes the results from zero to one, so that they are evenly apart. Although the resulting numbers are no longer comparable with any other version, I think it is now a lot more usable.

In this version every unique `kstwo()` result value in added to an array, which is then sorted and the smallest value gets 0.0 and the biggest 1.0 and everything in between proportionally based on their position on the array. Now the slider is a lot more useful, since it allows you to fine tune the results in higher detail. If you drag the slider to the far left, it shows only the charts with the lowest value (all of them have the same `kstwo()` result).

**Comment by andrewm4894** [comment]
> @andrewm4894 the latest commit (spread results evenly) re-distributes the results from zero to one, so that they are evenly apart. Although the resulting numbers are no longer comparable with any other version, I think it is now a lot more usable.

I think this is fine and was something I was expecting to have to do anyway once we move to cross node mc on the overview tab. 

The default for the slider on the FE will still be 0.05 and so end up now being a threshold ~ "at the top 5% most different" which should generally be fine. It also means you will always get results as opposed to edge case now where you could end up with no results if no kstwo results are less that the 0.05 pvalue.

So having this normalization of the results for ranking as input to the slider is useful and also let's us abstract the slider from the underlying algorithm (which was one thing I wanted to do originally).

I also like the idea of having multiple algos users can play with here and have a few more ideas with might look at down the line (PCA based reconstruction error of highlight area, Median Absolute Deviation, Earth Movers Distance, and maybe one or two more as well as Anomaly bit versions of both kstwo and volume methods, where you just run same Algo but on the anomaly bits instead of the raw data).

**Comment by andrewm4894** [comment]
@shyamvalsan if/when this gets merged we might need to think about adding a drop-down of some sort to the FE so the user can pick the algo. We can talk about this separate as a product discussion.

**Comment by ktsaou** [comment]
I think it is now ready to be merged.

I also added support for the options of the queries, so now we can also correlate anomaly rates!

**Comment by ktsaou** [comment]
Although, without a visual representation of the anomaly rates, it is a bit awkward. You can't understand why they are correlated...

**Comment by andrewm4894** [comment]
> Although, without a visual representation of the anomaly rates, it is a bit awkward. You can't understand why they are correlated...

...oh yeah :)

Maybe when Anomaly rate on every chart is available we might be able to plug into that in the results in some way if we decide to expose that option.

**Comment by shyamvalsan** [comment]
@ktsaou @andrewm4894 do we have results comparing the ks vs volume based approaches? we should collect some empirical evidence and pick one as the default based on accuracy and performance.

Its good to have a config option to choose different metrics correlation algorithm, **however**, from a user perspective I don't think we should offer multiple choices (via UI) as most users would just want to see correlated metrics without wanting to or needing to understand what goes on behind the curtain. 

And agree with @ilyam8 we definitely need UT that states what the ground truth is and how close we're getting to it.

**Comment by andrewm4894** [comment]
> @ktsaou @andrewm4894 do we have results comparing the ks vs volume based approaches? we should collect some empirical evidence and pick one as the default based on accuracy and performance.
> 

No and this is not possible as there are no labels and so no notion of accuracy

**Comment by andrewm4894** [comment]
> Its good to have a config option to choose different metrics correlation algorithm, **however**, from a user perspective I don't think we should offer multiple choices (via UI) as most users would just want to see correlated metrics without wanting to or needing to understand what goes on behind the curtain.
>

kstwo should be the default but we need to lean in more to giving users more options (those that want it) as there is never one size fits all and so being able to play with multiple algos if you wish easily and painlessly as part of using MC to explore is exactly the sort of thing we should be doing. Diminishing returns up.to a point but being able to try a handful of different algos would only make overall impact of MC more robust imo.

**Comment by shyamvalsan** [comment]
@andrewm4894 @ktsaou Volume (or average) based approach may not work depending on the distribution shape of the data, that's why I think we need more empirical evidence.

Could we not generate a bunch of metrics with varying levels of correlation (via a custom collector, maybe using something like https://scipy-cookbook.readthedocs.io/items/CorrelatedRandomSamples.html) and then see how closely MC is able to match the ground truth?

**Comment by andrewm4894** [comment]
> Could we not generate a bunch of metrics with varying levels of correlation

> @andrewm4894 @ktsaou Volume (or average) based approach may not work depending on the distribution shape of the data, that's why I think we need more empirical evidence.
> 
> Could we not generate a bunch of metrics with varying levels of correlation (via a custom collector, maybe using something like https://scipy-cookbook.readthedocs.io/items/CorrelatedRandomSamples.html) and then see how closely MC is able to match the ground truth?

A lot of work and what would that actually really "prove". 

I was thinking about just building a little streamlit app in python that would do both algos (and any else we wanted, emd, mad, pca) so that anyone could easily just play with them and qualitatively see if we like them or not.

**Comment by ktsaou** [comment]
Once this is merged, remember to pass `&options=raw` to get whatever the algorithm returns back, that is without the even distribution the post-processing does. The same parameter should be used when aggregating multiple nodes together.

For my taste the volume is more predicable for users (easier to understand and explain) and it is really a lot faster (just query time - nothing else).

## User controls

There are 3 controls supported currently:

1. the method: `ks2` or `volume`.
2. use the data or the anomaly rate (`&options=anomaly-bit`).
3. `&group=` function to aggregate the values over time: `average` (the default), or `median`, `min`, `max`, `sum`, `stddev`, `ses` or `ema` (simple exponential smoothing), `des` or `ewma` (double exponential smoothing), `cv` or `rsd` (coefficient of variation - like stddev but as %)

I think we should expose all these to the UI and use them to understand the value they provide. Once we see them, we can then hide the ones that are not that important.

## Current UI issues

The UI currently has the following problems:

1. The slider needs to be as wide as possible. All other controls should be minimal. The slider needs to get all the space it can, and expand with the width of the window to utilize all available space.
2. It would ideal if the slider could show the number of dimensions that will be matches which the user drags it.
3. If you close the MC panel, you lose the highlighted timeframe. This is stupid. The user needs the timeframe. It is useful for anomaly advisor, visual investigation, etc. We should never dismiss a highlighted window indirectly.
4. There are some cases on repeated use of the MC, where charts are shown with all the dimensions unselected! This should never happen.
5. Sometimes, moving the slider corrupts the display. Charts are shown on top of each other.
6. Moving to the bottom of the page and then making the slider more strict (moved to the left), leaves the user with a blank page. The user has to scroll up to see the charts.
7. The UI times out at about 6-7 seconds. The MC in this PR accepts a `&timeout=MS` option so that the agent will timeout, not the UI. The default in the agent is 60 seconds. For very long durations, a timeout of 6-7 seconds actually kills the benefits of MC.

## Metrics correlation across nodes

We need a backend service to do metric correlations across multiple nodes. The agents should be queried with `&options=raw` and the dataset they produce can be averaged across multiple nodes per context.

@shyamvalsan now is the time to do it. I can explain the algorithm for the aggregation in detail. It is very simple. A few lines of code.

**Comment by andrewm4894** [comment]
Just doing some testing - i have 3 nodes and running same load based on cronjobs and all created at the same time by terraform. 

- `staging-devml-ml-enabled-orphan` = mc is enabled will use current MC on agent.
- `staging-devml-ml-enabled-orphan-ktmc` = mc is enabled but its running this branch so usese this version of MC.
- `staging-devml-ml-enabled-orphan-ncmc` = mc is disabled and node is not streaming to parent so will just use cloud MC.

I selected a same highlight window on all 3 nodes and pressed the green button (slightly one after the other).

I do indeed see that the current agent version has worst impact here in terms of cpu and is very similar to the cloud (neon purple vs pink line) and that the agent MC of this branch `-ktmc` seems to have a much much lower impact that both others

image
image

Next I will do some functional / user acceptance tests of the actual results, but resource performance wise it looks like there are indeed significant gains here from what I can tell.

**Comment by andrewm4894** [comment]
[This](https://drive.google.com/file/d/1P9hUeZCPkIP8dEQVCifkqDLs8sZafdvy/view?usp=sharing) is just a reference artifact in case we ever want it later - a ~13 min video of me doing a qualitative test of the results. 

**tl;dr** results look good to me in terms of accuracy etc and i also see good evidence as to why this should actually be enabled by default on the agent since in a lot of cases will be less impactful than current MC on cloud.

**Comment by andrewm4894** [comment]
fyi - creating [this GH discussion](https://github.com/netdata/netdata/discussions/13117) topic for chat on cross node MC ideas so we can keep this PR cleanish.

**Comment by ktsaou** [comment]
@andrewm4894 I added unittests and made the necessary changes to be 100% accurate against the python `ks_2samp()`. The code had an issue and was producing wrong results under some conditions. Now it should be ok.

I changed the default to `volume` because I think it would be better for aggregating across nodes.

I also added a `netdata.conf` option if you want to change the default:

```
[global]
   enable metric correlations = yes
   metric correlations method = volume
```

The ks algorithm is called `ks2`.

**Comment by andrewm4894** [comment]
cool - ill drop and recreate my nodes to get latest commit.

I would probably not be so quick to use `volume` as default until we all dogfood it a bit more. I'll make two nodes one with each and will test it more.

**Comment by ktsaou** [comment]
rebased to master

**Review by ilyam8 — APPROVED** [review]

**Pull request merged**
